### ISSUE TITLE
.data does not work if object has same name as variable

### ISSUE THREAD

**Issue by sollano** [tidyverse/dplyr#2916]
I believe this is a bug with the ```.data``` pronoun:

```r
# Load package and example data
library(dplyr)

# this works

my_var <- "homeworld"

starwars %>%
  group_by(.data[[my_var]]) %>%
  summarise_at(vars(height:mass), mean, na.rm = TRUE)

# A tibble: 49 x 3
           my_var   height  mass
            <chr>    <dbl> <dbl>
 1       Alderaan 176.3333  64.0
 2    Aleen Minor  79.0000  15.0
 3         Bespin 175.0000  79.0
 4     Bestine IV 180.0000 110.0
 5 Cato Neimoidia 191.0000  90.0
 6          Cerea 198.0000  82.0
 7       Champala 196.0000   NaN
 8      Chandrila 150.0000   NaN
 9   Concord Dawn 183.0000  79.0
10       Corellia 175.0000  78.5
# ... with 39 more rows

# this does not

homeworld <- "homeworld"

starwars %>%
  group_by(.data[[homeworld]]) %>%
  summarise_at(vars(height:mass), mean, na.rm = TRUE)

Error in mutate_impl(.data, dots) : 
  Evaluation error: Must subset with a string.
```

**Comment by JohnMount** [comment]
I see the same problem, I would wonder if it is related to [`dplyr` issue 2904](https://github.com/tidyverse/dplyr/issues/2904) where similar "coincidence in naming" alias is happening (but didn't seem as unfavorable at first look).  I'd also say the odds of accidentally guessing a column name when coding are not actually that low.

**Comment by sollano** [comment]
@JohnMount You're right, the name of the grouping variable is modified to be equal to the  object's name. Really weird.
 Also, The guessing problem is really real to me, I usually use variable names that are initials of some measure, so it can be easily guessed.

**Comment by lionel-** [comment]
> Really weird.

That's the whole point of the dplyr interface though ;) We just need to make an exception for `.data` subsetting.

> The guessing problem is really real to me

As John mentioned, a general solution to avoid hitting variables from the data frame is to evaluate eagerly by unquoting with `!!`. E.g.:

```r
mass <- 100

transmute(starwars,
  height,
  eagerly = height * (!! mass),  # takes `mass` from environment
  lazily = height * mass         # takes `mass` from `starwars`
)
#> # A tibble: 87 x 3
#>    height eagerly lazily
#>     <int>   <dbl>  <dbl>
#>  1    172   17200  13244
#>  2    167   16700  12525
#>  3     96    9600   3072
#>  4    202   20200  27472
#>  5    150   15000   7350
#>  6    178   17800  21360
#>  7    165   16500  12375
#>  8     97    9700   3104
#>  9    183   18300  15372
#> 10    182   18200  14014
#> # ... with 77 more rows
```

**Comment by sollano** [comment]
> As John mentioned, a general solution to avoid hitting variables from the data frame is to evaluate eagerly by unquoting with !!.

Yes, that works when the object is a number. But If that object references a quoted variable name, and they have the same name, it doesn't, unless the object's name is the same as the variable name! (This is getting confusing to me). i.e:

``` r
this works:
height <- "height"
height <- quo(height)

print(height)

<quosure: global>
~height

transmute(starwars,
          height,
          height_sqrt = height * (!!height)         
)

# A tibble: 87 x 2
   height height_sqrt
    <int>       <int>
 1    172       29584
 2    167       27889
 3     96        9216
 4    202       40804
 5    150       22500
 6    178       31684
 7    165       27225
 8     97        9409
 9    183       33489
10    182       33124
# ... with 77 more rows

# this does not

h <- "height"
h <- quo(h)

print(h)

<quosure: global>
~h

transmute(starwars,
          height,
          height_sqrt = height * (!!h)         
)

Error in mutate_impl(.data, dots) : 
  Evaluation error: non-numeric argument to binary operator.

```
So this method does not work for functions either, because the arguments would have to match the names of the dataframe, otherwise it would fail. And generally, in a function, you want to use generic variable names, that can be replaced by the user.

So one is stuck with the .data pronoun, if one does want to use strings as input to his function. Maybe I'm getting something wrong here, but I could not make  !! convert strings to bare variable names.

**Comment by lionel-** [comment]
Let's look at what happens in this snippet:

```r
height <- "height"    # height refers to a column name
height <- quo(height) # height is a quosure referring to `height`, which is itself
```

`height` contains a quote to itself, so you won't do anything useful with it. Except if the symbol `height` is also contained in the data frame, then `height` column will prevail because scoping in dplyr is always dataframe → environment, with or without quosures.

So when you're doing this:

```r
h <- "height" # `h` contains a column name
h <- quo(h)   # `h` contains a quote of itself
```

Unquoting `h` refers to `h`, which isn't a column name but is found in the environment, and that `h` contains a reference to itself.

What you want is simply this:

```r
# Import `sym()` to create symbols
# will soon be exported in dplyr as well
sym <- rlang::sym

height_name <- "height"
height_sym <- sym(height_name)

transmute(starwars,
   height,
   height_sqrt = height * (!! height_sym)
)
```

this is equivalent to:

```r
height <- "height" # storing column name

transmute(starwars,
   height,
   height_sqrt = height * (!! sym(height)) # creating and inlining symbol from stored column name
)
```

Try to wrap with `expr()` to debug what expression dplyr sees:

```r
expr(transmute(starwars,
   height,
   height_sqrt = height * (!! sym(height))
))
```

**Comment by sollano** [comment]
@lionel- Thanks for that man, I really appreciate it. Just tested this in my function, and it works. When this sym function gets exported to dplyr, I think it would be a good idea to add an example like this to the programming vignette.  

But this makes me wonder... Doesn't this sym function kinda replaces the .data pronoun?

**Comment by lionel-** [comment]
> Doesn't this sym function kinda replaces the .data pronoun?

No because the `.data` pronoun ensures the variable comes from the data frame and never from the environment. Since we have the `df → env` scoping in dplyr, there is always this uncertainty about where the symbols are found.

Most of the time it doesn't matter (otherwise your dplyr pipelines wouldn't work) but when you want to be really explicit you have two solutions: unquoting eagerly makes sure you're taking your variable from the environment, and using the `.data` pronoun makes sure you're taking your variable from the data frame.

**Comment by sollano** [comment]
Oh, now I see.  It would be wiser to use the```.data``` pronoun, just to be sure.

**Comment by krlmlr** [comment]
Can we close this? Do we need to improve documentation?

**Comment by sollano** [comment]
I'm still kinda of confused with this because `rlang::sym` and `rlang::syms` are doing the jobs of `dplyr::quo` and `dplyr::quos`, so I simply don't use them anymore. I believe this should be detailed in the documentation. Most people don't even know about those rlang functions. That's just my opinion anyway, maybe I got something wrong. But even if I did, that just proves my point.

**Comment by lionel-** [comment]
> rlang::sym and rlang::syms are doing the jobs of dplyr::quo and dplyr::quos, just fine for me, so I simply don't use them anymore

Probably because you use them to refer to objects from the data frame. In this case they are completely fine. However in more complex cases (i.e. involving calls to functions) you'll need quosures to avoid bugs.

> believe this should be detailed in the documentation.

Better documentation is forthcoming.

**Comment by sollano** [comment]
>  However in more complex cases (i.e. involving calls to functions) you'll need quosures to avoid bugs.
Oh, thanks. I'll take note on this.

So the issue that I found is not really a bug, it simply happens because of the way quosures works, right? I think this is solved, then.

**Comment by krlmlr** [comment]
Keeping open, feel free to close when docs are updated.

**Comment by lionel-** [comment]
Upstream issue: https://github.com/r-lib/rlang/issues/477

**Comment by lionel-** [comment]
This is fixed with rlang 0.3.0 (which will be on CRAN in 1 or 2 weeks) where `.data[[` is now an unquote operator. This way the subsetting variable is never masked by data.
